### PR TITLE
Improve --compile-all error message and a round of verifier cleanups.

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

**Comment by lambdageek** [comment]
@kumpera, high-level question before I start reviewing in detail:
So I see that the old verifier interfaces had a `GSList **errors` outparam.  So presumably you could signal that multiple things were wrong.  Now there's just a single `MonoError *error` outparam which usually means that we bail as soon as one thing goes wrong.

Did the verifier never make use of the error list or did you change it now to only return the first error?

File: mono/metadata/metadata-verify.c [hunk 1]
```diff
@@ -4238,51 +4261,59 @@ verify_user_string (VerifyContext *ctx, guint32 offset)
...
 	ctx.stage = STAGE_TABLES;
 
 	is_valid_cattr_blob (&ctx, offset);
 
-	return cleanup_context (&ctx, error_list);
+	return cleanup_context_checked (&ctx, error);
 }
 
+
+//FIXME this should raise a System.Reflection.CustomAttributeFormatException
```
Comment (lambdageek) [inline mono/metadata/metadata-verify.c]: I really hope this change triggers an SRE regression test failure. 😞

**Comment by marek-safar** [comment]
@kumpera great patch, could you spent some more time on finishing touches (it causes stack overflow during build)

File: mono/metadata/image.c [hunk 2]
```diff
@@ -1299,11 +1299,13 @@ static MonoImage *
 do_mono_image_load (MonoImage *image, MonoImageOpenStatus *status,
 		    gboolean care_about_cli, gboolean care_about_pecoff)
 {
+	ERROR_DECL (error);
 	MonoCLIImageInfo *iinfo;
 	MonoDotNetHeader *header;
-	GSList *errors = NULL;
 	GSList *l;
 
+	error_init (error);
```
Comment (luhenry) [inline mono/metadata/image.c]: This is already done by `ERROR_DECL`

File: tools/pedump/pedump.c [hunk 3]
```diff
@@ -422,15 +423,15 @@ usage (void)
...
 	MonoImage *image;
 	MonoTableInfo *table;
 	MonoAssembly *assembly;
 	MonoImageOpenStatus status;
 	int i, count = 0;
-	const char* desc [] = {
-		"Ok", "Error", "Warning", NULL, "CLS", NULL, NULL, NULL, "Not Verifiable"
-	};
+
+	error_init (error);
```
Comment (luhenry) [inline tools/pedump/pedump.c]: This is already done by `ERROR_DECL`

**Comment by marek-safar** [comment]
@kumpera any plans to finish this PR?

**Comment by kumpera** [comment]
@marek-safar yes, I need to fix the regression it introduces first.

**Comment by kumpera** [comment]
@lambdageek the functions I changed don't need an error list as their callers needed to set a MonoError in the end.

The list is a side-effect of the original verifier designer that was meant to be used offline and report as much as possible, instead of fail-fast.

**Comment by luhenry** [comment]
@monojenkins squash

**Comment by luhenry** [comment]
@monojenkins build failed

**Comment by luhenry** [comment]
@kumpera it fails to build.

**Comment by kumpera** [comment]
then let me refresh this on top of master

**Pull request merged**